File: src/termino/storage.py
"""Session logging for Termino.

All sessions are stored as JSON in ~/.termino/sessions.json.
"""

import json
import os
from datetime import datetime, timezone
from typing import Optional


DATA_DIR = os.path.expanduser("~/.termino")
SESSIONS_FILE = os.path.join(DATA_DIR, "sessions.json")


def _ensure_data_dir() -> None:
    """Create the data directory if it doesn't exist."""
    os.makedirs(DATA_DIR, exist_ok=True)


def _load_sessions() -> list:
    """Load all sessions from the sessions file.

    Returns:
        List of session dicts.
    """
    _ensure_data_dir()
    if not os.path.exists(SESSIONS_FILE):
        return []
    try:
        with open(SESSIONS_FILE, "r") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError):
        return []


def _save_sessions(sessions: list) -> None:
    """Save sessions to the sessions file.

    Args:
        sessions: List of session dicts.
    """
    _ensure_data_dir()
    with open(SESSIONS_FILE, "w") as f:
        json.dump(sessions, f, indent=2)


def save_session(
    session_type: str,
    duration_seconds: float,
    status: str = "completed",
    laps: Optional[list] = None,
    cycles: Optional[list] = None,
) -> dict:
    """Save a session to the session log.

    Args:
        session_type: Type of session (stopwatch, countdown, pomodoro).
        duration_seconds: Total duration in seconds.
        status: Session status (completed, interrupted).
        laps: List of lap dicts (for stopwatch).
        cycles: List of cycle dicts (for pomodoro).

    Returns:
        The saved session dict.
    """
    now = datetime.now(timezone.utc).isoformat()
    session = {
        "type": session_type,
        "started": now,
        "ended": now,
        "duration_seconds": round(duration_seconds, 3),
        "status": status,
    }
    if laps:
        session["laps"] = [
            {"lap": l["lap"], "split": round(l["split"], 3), "total": round(l["total"], 3)}
            for l in laps
        ]
    if cycles:
        session["cycles"] = [
            {"cycle": c["cycle"], "type": c["type"], "duration": round(c["duration"], 3)}
            for c in cycles
        ]

    sessions = _load_sessions()
    sessions.append(session)
    _save_sessions(sessions)
    return session


def get_sessions(session_type: Optional[str] = None, limit: Optional[int] = None) -> list:
    """Get sessions, optionally filtered by type and limited.

    Args:
        session_type: Optional filter by session type.
        limit: Optional max number of sessions to return (most recent first).

    Returns:
        List of session dicts.
    """
    sessions = _load_sessions()
    if session_type:
        sessions = [s for s in sessions if s.get("type") == session_type]
    # Reverse so most recent first
    sessions.reverse()
    if limit is not None:
        sessions = sessions[:limit]
    return sessions


def get_session_count() -> int:
    """Get the total number of logged sessions.

    Returns:
        Session count.
    """
    return len(_load_sessions())

File: src/storage.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

/// A session lap entry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionLap {
    pub lap: u32,
    pub split: f64,
    pub total: f64,
}

/// A pomodoro cycle entry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionCycle {
    pub cycle: u32,
    #[serde(rename = "type")]
    pub cycle_type: String,
    pub duration: f64,
}

/// A recorded session.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionData {
    #[serde(rename = "type")]
    pub session_type: String,
    pub started: String,
    pub ended: String,
    pub duration_seconds: f64,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub laps: Option<Vec<SessionLap>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cycles: Option<Vec<SessionCycle>>,
}

/// Get the data directory path.
fn data_dir() -> PathBuf {
    if let Ok(home) = std::env::var("TERMINO_HOME") {
        return PathBuf::from(home);
    }
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(".termino")
}

/// Get the sessions file path.
fn sessions_file() -> PathBuf {
    data_dir().join("sessions.json")
}

/// Load all sessions from file.
fn load_sessions() -> Vec<SessionData> {
    let path = sessions_file();
    if !path.exists() {
        return Vec::new();
    }
    match fs::read_to_string(&path) {
        Ok(content) => serde_json::from_str(&content).unwrap_or_default(),
        Err(_) => Vec::new(),
    }
}

/// Save sessions to file.
fn save_sessions(sessions: &[SessionData]) -> anyhow::Result<()> {
    let dir = data_dir();
    fs::create_dir_all(&dir)?;
    let content = serde_json::to_string_pretty(sessions)?;
    // Atomic write: write to temp, then rename
    let tmp_path = dir.join("sessions.json.tmp");
    fs::write(&tmp_path, &content)?;
    fs::rename(&tmp_path, sessions_file())?;
    Ok(())
}

/// Save a session to the log.
pub fn save_session(session: &SessionData) -> anyhow::Result<SessionData> {
    let mut sessions = load_sessions();
    sessions.push(session.clone());
    save_sessions(&sessions)?;
    Ok(session.clone())
}

/// Get sessions, optionally filtered by type and limited.
pub fn get_sessions(
    session_type: Option<&str>,
    limit: Option<usize>,
) -> anyhow::Result<Vec<SessionData>> {
    let mut sessions = load_sessions();
    if let Some(st) = session_type {
        sessions.retain(|s| s.session_type == st);
    }
    // Reverse so most recent first
    sessions.reverse();
    if let Some(limit) = limit {
        sessions.truncate(limit);
    }
    Ok(sessions)
}

/// Get total session count.
pub fn get_session_count() -> usize {
    load_sessions().len()
}
